### PR TITLE
Clean up for Debian

### PR DESCRIPTION
**Pull request by h4ck3rm1k3**
Leave the deleting of alternatives to the OS, don't just delete the
file.
Also resolve the symlink to check it. If this is installed, you don't
need to do anything.
- providers/alternatives.rb:
- recipes/default_java_symlink.rb

### CODE REVIEW

File: providers/alternatives.rb [hunk 1]
```diff
@@ -38,8 +38,9 @@
         description = "Add alternative for #{cmd}"
         converge_by(description) do
           Chef::Log.debug "Adding alternative for #{cmd}"
-          shell_out("rm /var/lib/alternatives/#{cmd}")
-          install_cmd = shell_out("#{alternatives_cmd} --install #{bin_path} #{cmd} #{alt_path} #{priority}")
+          Chef::Log.debug "Runnning #{alternatives_cmd} --install #{bin_path} #{cmd} #{alt_path} #{priority}"
+
+          install_cmd = shell_out("sudo #{alternatives_cmd} --install #{bin_path} #{cmd} #{alt_path} #{priority}")
```
Comment (erichelgeson) [inline providers/alternatives.rb]: This will be run as root, so no need to sudo.

File: providers/alternatives.rb [hunk 2]
```diff
@@ -38,8 +38,9 @@
         description = "Add alternative for #{cmd}"
         converge_by(description) do
           Chef::Log.debug "Adding alternative for #{cmd}"
-          shell_out("rm /var/lib/alternatives/#{cmd}")
```
Comment (erichelgeson) [inline providers/alternatives.rb]: When running alternatives commands, you mention the OS will clean up, what do you mean by that? Wont you have orphaned alternative entries if you dont rm it?

File: recipes/default_java_symlink.rb [hunk 3]
```diff
@@ -12,8 +12,24 @@
 # WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 # See the License for the specific language governing permissions and
 # limitations under the License.
+require 'pathname'
 
-link '/usr/lib/jvm/default-java' do
-  to node['java']['java_home']
-  not_if { node['java']['java_home'] == '/usr/lib/jvm/default-java' }
```
Comment (erichelgeson) [inline recipes/default_java_symlink.rb]: Doesnt this not if do what all the code below does anyways? Or if theres a different check we can add to the guard to do what you're trying to do below?

**Comment by h4ck3rm1k3** [comment]
Well in my case this is installed, and why would I remove it first, in fact I am running as a non root user and dont have the permissions to remove the link that will be added. http://linux.die.net/man/8/update-alternatives will remove the old link if needed. 

Now for the symlink check the code is definitely not finding the current file and trying to redo it. You need to resolve the symlink, and on debian is it is a relative one. That means the string compare will fail.

**Comment by erichelgeson** [comment]
I have to assume chef is running as root, its an edge case to not. You'd have to run chef as a user with NOPASSWD sudo anyways to make this work, which is just going down the rabbit hole :)

I'll play around with the symlink code a bit and see if I can reproduce that issue on debian.

File: providers/alternatives.rb (hunk 1 above)
Comment (h4ck3rm1k3) [inline providers/alternatives.rb]: yes, ooops I thought that got removed. I am not running this as root. I am testing it out and executing individual commands manually to see what it does.

File: providers/alternatives.rb (hunk 2 above)
Comment (h4ck3rm1k3) [inline providers/alternatives.rb]: you are removing what is there, but debian manages this all for you. http://linux.die.net/man/8/update-alternatives

File: recipes/default_java_symlink.rb (hunk 3 above)
Comment (h4ck3rm1k3) [inline recipes/default_java_symlink.rb]: Pathname.new(default).realpath that is the real key here. You need it.

**Comment by h4ck3rm1k3** [comment]
Here is what this looks like

    ls -latr /usr/lib/jvm/default-java
    lrwxrwxrwx 1 root root 24 Aug  5  2013 /usr/lib/jvm/default-java -> java-1.7.0-openjdk-amd64

you can see that is is a relative symlink

Package: openjdk-7-jre-headless
Version: 7u65-2.5.2-4

my node and test case is here : https://github.com/h4ck3rm1k3/chef-test-frame

the node file is here : 
https://github.com/h4ck3rm1k3/chef-test-frame/blob/master/node.json#L11

it does not recognise my arch either so I needed to set 

    "java_home" : "/usr/lib/jvm/java-7-openjdk-amd64",

otherwise it produces 

    "java_home" : "/usr/lib/jvm/java-7-openjdk" 

automatically

**Comment by bitmonk** [comment]
Aside from the sudo/non-root-user concern, which I agree shouldn't be supported, I think I agree with the intent of this, primarily that update-alternatives should be used directly, and not imitated with ruby code in the chef recipe.

Can I help move this along?

**Comment by erichelgeson** [comment]
Closing due to age, feel free to re-open with a updated PR addressing comments.

**Comment by h4ck3rm1k3** [comment]
I am not using chef any more, ansible is much simpler.

**Pull request closed without merging**